### DISCUSSION TITLE
Add typing hints?

### DISCUSSION THREAD

**Discussion by igo95862** [pyudev/pyudev#451] [Ideas]
https://docs.python.org/3/library/typing.html

Type hints allow static checkers to verify code and LSP servers to do completions.

It would also be possible to use static type checker such as `mypy` as a part of Continuous Integration.

I have an example of import C function and applying type hints in my project:

https://github.com/igo95862/bubblejail/blob/1ec598fb8f752ba89db58206113b9c3127214059/bubblejail/bubblejail_seccomp.py#L32

**Comment by mulkieran** [comment]
I'm not against type hints; but last I checked they didn't seem to be very effective in the code that I tried modifying. If you were interested in submitting a PR, you should 1) pick a spot you care about in the code, rather than attempting to annotate all the code, 2) introduce _some_ sort of testing of the typing into the CI. What sort of testing would be your choice, I haven't looked into what's available in a long time.

**Reply by igo95862** [reply]
I recommend this blogpost: https://sethmlarson.dev/blog/tests-arent-enough-case-study-after-adding-types-to-urllib3

TLDR: type hints > 100% test coverage

**Comment by mulkieran** [comment]
I don't develop this library, I maintain it. You, or someone else, must submit a PR. If that happens, I'll certainly consider it. If the typing ecosystem has improved since the last time I looked seriously[1], I'll be very glad.

[1] https://github.com/python/typeshed/pull/127

**Reply by igo95862** [reply]
Duh. It was in 2016. A lot has changed since then.

I can write a PR. What is the minimum Python version? The version 3.7  would be better because is has `from __future__ import annotations` or the type hints might not be compatible with Python 3.11. I see your Github Actions test against Python 3.6 but that Python version is no longer supported.

**Reply by mulkieran** [reply]
@igo95862 My Ph. D. thesis was in type theory; as a consequence I know quite a bit about types. This may be just the tone you adopt to everybody, but to me it comes across as insulting. Please moderate your tone and make an effort to be more polite, or I will have to block you entirely.

**Comment by igo95862** [comment]
I opened first Pull Request: https://github.com/pyudev/pyudev/pull/452

I started with leaf packages such as `pydev.os._pipe` because I can update them one by one and slowly add type hints to entire tree.

There is a lot of legacy code that would needs to be cleaned up. If you are not comfortable with reworking large parts of the code I can create a new package that is built on modern python idioms from the start.

**Reply by mulkieran** [reply]
I think you should make your own package! Best of luck.

**Reply by mulkieran** [reply]
I will review your PR; if you are still interested in having it accepted, but I expect you'll have to divide it into parts.
